algebra/free_group.lean

```diff
@@ -0,0 +1,423 @@
+sum T T
+
+namespace to_inv_type
+
+def inv (T : Type u) : to_inv_type T → to_inv_type T
+| (sum.inl x) := sum.inr x
+| (sum.inr x) := sum.inl x
+
+theorem inv.inv (T : Type u) : ∀ x : to_inv_type T, inv T (inv T x) = x
+| (sum.inl x) := rfl
+| (sum.inr x) := rfl
+
+instance (T : Type u) : inv_type (to_inv_type T) :=
+{ inv := inv T,
+  inv_inv := inv.inv T }
+
+def of_inv_type {T : Type u} : T → to_inv_type T :=
+sum.inl
+
+section left_adjoint
+
+parameters (T : Type u)
+parameters (IT : Type v) [inv_type IT]
+parameters (f : T → IT) -- T → Forgetful(IT)
+
+def left_adjoint : to_inv_type T → IT -- Free(T) → IT
+| (sum.inl x) := f x
+| (sum.inr x) := (f x)⁻¹
+
+theorem left_adjoint.inv : ∀ x, left_adjoint x⁻¹ = (left_adjoint x)⁻¹
+| (sum.inl x) := rfl
+| (sum.inr x) := eq.symm $ inv_type.inv_inv _
+
+theorem left_adjoint.commutes (x) : left_adjoint (of_inv_type x) = f x := rfl
+
+parameters (g : to_inv_type T → IT)
+parameters (Hg1 : ∀ x, g x⁻¹ = (g x)⁻¹)
+parameters (Hg2 : ∀ x, g (of_inv_type x) = f x)
+
+theorem left_adjoint.unique : ∀ x, g x = left_adjoint x
+| (sum.inl x) := Hg2 x
+| (sum.inr x) := (Hg1 (sum.inl x)).trans $ congr_arg _ $ Hg2 x
+
+end left_adjoint
+
+end to_inv_type
+
+@[reducible] def free_group.word (S : Type u) : Type u :=
+inv_type.to_inv_mon $ to_inv_type S
+
+@[reducible] def free_group (S : Type u) : Type u :=
+inv_mon.to_group $ free_group.word S
+
+namespace free_group
+
+variables (S : Type u)
+
+instance : group (free_group S) :=
+inv_mon.to_group.group _
+
+def of_type : S → free_group S :=
+λ x, ⟦[sum.inl x]⟧
+
+variables (G : Type v) [group G]
+variables (f : S → G) -- S → Forgetful(G)
+
+def to_group : free_group S → G := -- Free(S) → G
+λ x, inv_mon.to_group.left_adjoint _ G
+  (λ y, inv_type.to_inv_mon.left_adjoint (to_inv_type S) _
+    (to_inv_type.left_adjoint S _ f) y)
+  (inv_type.to_inv_mon.left_adjoint.mul _ _ _)
+  (inv_type.to_inv_mon.left_adjoint.one _ _ _)
+  (inv_type.to_inv_mon.left_adjoint.inv _ _ _
+    (to_inv_type.left_adjoint.inv _ _ _))
+  x
+
+def to_group.is_group_hom : is_group_hom (to_group S G f) :=
+inv_mon.to_group.left_adjoint.is_group_hom _ _ _ _ _ _
+
+def to_group.commutes (x) : to_group S G f (of_type S x) = f x :=
+mul_one _
+
+variables (g : free_group S → G)
+variables (Hg1 : is_group_hom g)
+variables (Hg2 : ∀ x, g (of_type S x) = f x)
+include Hg1 Hg2
+
+def to_group.unique : ∀ x, g x = to_group S G f x :=
+λ x, quotient.induction_on x $ λ m, list.rec_on m Hg1.one $
+λ h t ih, sum.rec_on h
+  (λ a, show g (of_type S a * ⟦t⟧) = _ , by rw [Hg1, Hg2, ih]; refl)
+  (λ a, show g ((of_type S a)⁻¹ * ⟦t⟧) = _ , by rw [Hg1, ← Hg1.inv, Hg2, ih]; refl)
+
+omit Hg1 Hg2
+
+variable [decidable_eq S]
+
+def reduce : word S → word S
+| ((sum.inl x)::(sum.inr y)::t) := if x = y then reduce t else ((sum.inl x)::(sum.inr y)::reduce t)
+| ((sum.inr x)::(sum.inl y)::t) := if x = y then reduce t else ((sum.inr x)::(sum.inl y)::reduce t)
+| (h1::h2::t)                   := h1::h2::reduce t
```

Comment:
This looks wrong to me, shouldn't `reduce [a, b, b⁻¹]` evaluate to `[a]`? As far as I see it this is not the case.